controller/src/main/java/org/jboss/as/controller/ServiceTarget.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2023 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jboss.as.controller;
+
+import org.jboss.msc.service.LifecycleListener;
+
+/**
+ * {@link org.jboss.msc.service.ServiceTarget} whose builders support capability requirements.
+ * @author Paul Ferraro
+ */
+public interface ServiceTarget extends org.jboss.msc.service.ServiceTarget {
```

Comment:
This interface is practically a kind of mirror of `CapabilityServiceTarget`. I'm wondering if we should try to avoid these name clashes introduced via this interface (I mean `ServiceTarget` vs `org.jboss.msc.service.ServiceTarget`), and consequently the binary incompatible issue, and try to find a solution by just adding `CapabilityServiceTarget getCapabilityServiceTarget()` in `DeploymentPhaseContext`. I don't know if that is possible or whether it was already checked, so this is just a comment.

This seems to be aligned with what `OperationContext` does. There we have deprecated the uses of `ServiceTarget getServiceTarget()` in favor of `CapabilityServiceTarget getCapabilityServiceTarget()`.

Would that make sense?

Reply:
> This interface is practically a kind of mirror of `CapabilityServiceTarget`.

No. It is a generalization of CapabilityServiceTarget.  CapabilityServiceTarget is used to create services for RuntimeCapability instances.  This new interface is used to create standard MSC services.

> I'm wondering if we should try to avoid these name clashes introduced via this interface (I mean `ServiceTarget` vs `org.jboss.msc.service.ServiceTarget`)

Why?

> ... , and consequently the binary incompatible issue.

The binary compatibility issue is not a consequence of the name clash of this interface with org.jboss.msc.service.ServiceTarget; it is a consequence of narrowing the return type of a public interface method (regardless of its name).

> ... and try to find a solution by just adding `CapabilityServiceTarget getCapabilityServiceTarget()` in `DeploymentPhaseContext`.

This would be bad interface design, since the object returned by DeploymentPhaseContext.getServiceTarget() cannot implement all of the methods of CapabilityServiceTarget.  More specifically, this object cannot create objects that implement all of the methods of CapabilityServiceBuilder, since this new object explicitly supports only standard MSC service installation.

Frankly, the question presented in the PR description needs to be answered first, i.e. are the deployment-related interfaces in org.wildfly.core:wildfly-server constrained by binary compatibility?

> This seems to be aligned with what `OperationContext` does. There we have deprecated the uses of `ServiceTarget getServiceTarget()` in favor of `CapabilityServiceTarget getCapabilityServiceTarget()`.
> Would that make sense?

If we decide that binary compatibility is a requirement for interfaces within the wildfly-server module, then, yes, adding a new method is essential.  However, we would still need the same generalization hierarchy proposed by this PR, i.e. CapabilityServiceTarget -> FooServiceTarget -> org.jboss.msc.service.ServiceTarget, and figure out a better name.

However, if we decide that binary compatibility is _not_  a requirement for the interfaces within the wildfly-server module (the case for which I am arguing), then we should avoid adding a new DeploymentPhaseContext.getFooServiceTarget() method and should prefer a source-compatible approach, such as what is being proposed by this PR.

Reply:
> In any case, these changes for sure need more eyes on it, since I'm not fully familiar with these interfaces and compositions.

Yes. I am hoping @bstansberry can provide his perspective on the binary-compatibility question -and that @ropalka can provide his MSC perspective.

Reply:
> No. It is a generalization of CapabilityServiceTarget. CapabilityServiceTarget is used to create services for RuntimeCapability instances. This new interface is used to create standard MSC services.

Yes right, CapabilityServiceTarget is used to create services for RuntimeCapability, but it also extends org.jboss.msc.service.ServiceTarget. So it looked to me quite similar to your new interface.


> Why?

No technical reasons, but if the two interfaces add different things, I would think that adding them with different names could make sense. Not a strong reason. I understand that creating org.jboss.as.controller.ServiceTarget, will allow us to avoid refactoring the callers; they can continue using the parent interface org.jboss.msc.service.ServiceTarget, but I got the impression that was going to be more convoluted than standardizing the CapabilityServiceTarget usage if possible.

OperationContext deprecated the uses of org.jboss.msc.service.ServiceTarget, which also made me think we wanted to eliminate the ServiceTarget usages. However we are now adding a new ServiceTarget interface, yes an interface in a new package, but I'm more talking about what I thought there were the future intentions: I assumed CapabilityServiceTarget in favor of ServiceTarget.

Reply:
> Yes right, CapabilityServiceTarget is used to create services for RuntimeCapability, but it also extends org.jboss.msc.service.ServiceTarget. So it looked to me quite similar to your new interface.

This change makes CapabilityServiceTarget a _specialization_ of the new interface.

> No technical reasons, but if the two interfaces add different things, I would think that adding them with different names could make sense.

That's the point.  They produce the same thing, but WildFly's ServiceTarget returns a ServiceBuilder specialization that adds resolution of capability requirements.

> OperationContext deprecated the uses of org.jboss.msc.service.ServiceTarget, which also made me think we wanted to eliminate the ServiceTarget usages.

This was done because the two methods were redundant. OperationContext.getServiceTarget() and Operation.getCapabilityServiceTarget() returned the exact same object.

> However we are now adding a new ServiceTarget interface, yes an interface in a new package, but I'm more talking about what I thought there were the future intentions: I assumed CapabilityServiceTarget in favor of ServiceTarget.

AIUI, CapabilityServiceTarget is not appropriate for use within a DeploymentPhaseContext.  From what I can tell, the concept of a RuntimeCapability is only really relevant to the server management model, not to a deployment management model, where there is no such thing as deployment model validation independent of runtime.  Thus, a DeploymentUnitProcessor would have no use for the complete functionality of CapabilityServiceTarget.  @bstansberry Am I mistaken?  If so, then you are right, and we could just expose CapabilityServiceTarget from DeploymentPhaseContext.

Reply:
@pferraro You are correct; a deployment cannot provide a capability. The deployment resource definition cannot register at server boot that the resource *could* provide a capability, and when the concrete resource instance is added the add op cannot register at Stage.MODEL the fact that the deployment provides the capability. So all the deployment could do at runtime would be to install a service under the name of a capability registered by some 'foreign' resource. Which would be wrong. Perhaps if some day we implemented the old idea of a 'system deployment' this could somehow be something such a thing does, but that's pure speculation.

Re source and binary compatibility, IMHO most packages in the wildfly-controller module and the wildfly-server module's org.jboss.as.server.deployment package tree should *try very hard* to maintain binary compatibility but are only *required* to maintain source compatibility. Except for removal of long-deprecated API. Binary compatibility breakage should be infrequent. Ideally we batch such breakage up.

Binary incompatibility in these packages forces external extensions to rebuild and release, and perhaps to branch.

Reply:
> most packages in the wildfly-controller module and the wildfly-server module's org.jboss.as.server.deployment package tree should try very hard to maintain binary compatibility

@bstansberry ok - I will try harder... :)  IOW, I'll leave the existing getServiceTarget() intact and add a separate method.

Reply:
@yersan @bstansberry  OK - I renamed these interfaces accordingly:
org.jboss.as.controller.ServiceTarget -> RequirementServiceTarget
org.jboss.as.controller.ServiceBuilder -> RequirementServiceBuilder

I restored the original DeploymentPhaseContext.getServiceTarget() to return its original type and added a new DeploymentPhaseContext.getRequirementServiceTarget() method.
Feel free to suggest better names for these interfaces.